File: Graph/building-roads.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <unordered_map>

using namespace std;

void bfs(vector<vector<int>>& adj, vector<int>& visited, int start) {
        queue<int> q;
        q.push(start);
        visited[start] = 1;

        while (!q.empty()) {
            int node = q.front();
            q.pop();

            for (int neighbor : adj[node]) {
                if (!visited[neighbor]) {
                    q.push(neighbor);
                    visited[neighbor] = 1;
                }
            }
        }
    }

int main() {
    int n, m;
    cin >> n >> m;

    vector<vector<int>> adj(n + 1); // Adjacency list representation
    vector<int> visited(n + 1, 0);   // Visited array to keep track of visited nodes

    // Assuming the graph is given as edges
    for (int i = 0; i < m; ++i) {
        int u, v;
        cin >> u >> v;
        adj[u].push_back(v);
        adj[v].push_back(u);
    }

    vector<int> comps;
    queue<int> q;

    for (int i = 1; i <= n; ++i) {
        if (!visited[i]) {
            bfs(adj, visited, i);
            comps.push_back(i);
        }
    }

    if (comps.size() <= 1) {
        cout << 0 << endl;
    } else {
        cout << comps.size() - 1 << endl;
        for (int i = 1; i < comps.size(); ++i) {
            cout << comps[i] << ' ' << comps[i - 1] << endl;
        }
    }

    return 0;
}

File: Graph/building-roads.py
from collections import defaultdict

n, m = map(int, input().split())
arr = [[int(x) for x in input().split()] for _ in range(m)]

adj = defaultdict(list)

for a1, a2 in arr:
    adj[a1].append(a2)
    adj[a2].append(a1)

comps = []

vis = [0] * (n+1)
q = []

# def dfs(node):
#     vis[node] = 1

#     for it in adj[node]:
#         if vis[it]==0:
#             dfs(it)

#     return

def bfs(node, q):
    q.append(node)
    while len(q)>0:
        popped = q.pop(0)   
        vis[popped] = 1
        for it in adj[popped]:
            if vis[it] == 0:
                q.append(it)

    return

for i in range(1, n+1):
    if vis[i] == 0:
        q = []
        bfs(i, q)
        comps.append(i)

if len(comps) <= 1:
    print(0)
else:
    print(len(comps)-1)
    for i in range(1, len(comps)):
        print(str(comps[i]) + ' ' + str(comps[i-1]))



